### PR TITLE
refactor update-api-reference-docs.sh

### LINKED ISSUE #20877: update-api-reference-docs needs to do all groups
It's not obvious how to add more groups to this file. This blocks job & hpa moves.

https://github.com/kubernetes/kubernetes/blob/master/hack/update-api-reference-docs.sh#L43

@nikhiljindal or @caesarxuchao, this file looks like yours. Can one of you make it take a list of groups like other similar files?

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #20877.

### CODE REVIEW

File: hack/update-api-reference-docs.sh [hunk 1]
```diff
@@ -56,13 +57,25 @@ if [[ $(uname) == "Darwin" ]]; then
...
-    https://raw.githubusercontent.com/kubernetes/kubernetes/master/pkg/apis/extensions/v1beta1/register.go
+for ver in $VERSIONS; do
+  TMP_IN_HOST="${OUTPUT_TMP_IN_HOST}/${ver}"
+  REGISTER_FILE_URL="https://raw.githubusercontent.com/kubernetes/kubernetes/master/pkg"
+  if [[ ${ver} == "v1" ]]; then
+    REGISTER_FILE_URL="${REGISTER_FILE_URL}/api/${ver}/register.go"
+  else
+    REGISTER_FILE_URL="${REGISTER_FILE_URL}/apis/${ver}/register.go"
+  fi
+  # TODO: we need to rename the swagger JSON files, currently they are named as
```
Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: We need to address this TODO. I'll do it in a follow up PR.

Comment (lavalamp) [inline hack/update-api-reference-docs.sh]: Yes, this is necessary, since batch is starting out in v1. Is it possible to do it in this PR?

**Comment by lavalamp** [comment]
Thanks, this looks much better!

File: hack/update-api-reference-docs.sh (hunk 1 above)
Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: Yeah, I'm just trying to make the review easier. I'll add another commit.

Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: New commit pushed. PTAL.

File: hack/after-build/update-swagger-spec.sh [hunk 2]
```diff
@@ -64,15 +64,36 @@ APISERVER_PID=$!
...
+for ver in ${VERSIONS}; do
+  # fetch the swagger spec for each group version. 
+  if [[ ${ver} == "v1" ]]; then
+    SUBPATH="api"
+  else
+    SUBPATH="apis"
+  fi
+  SUBPATH="${SUBPATH}/${ver}"
+  SWAGGER_JSON_NAME="$(kube::util::gv-to-swagger-name ${ver}).json"
+  curl -fs ${SWAGGER_API_PATH}${SUBPATH} > ${SWAGGER_ROOT_DIR}/${SWAGGER_JSON_NAME}
```
Comment (krousey) [inline hack/after-build/update-swagger-spec.sh]: These should probably all be double quoted.

File: hack/after-build/update-swagger-spec.sh [hunk 3]
```diff
@@ -64,15 +64,36 @@ APISERVER_PID=$!
...
+  
+  # fetch the swagger spec for the discovery mechanism at group level.
+  if [[ ${ver} == "v1" ]]; then
+    continue 
+  fi
+  SUBPATH="apis/"${ver%/*}
+  SWAGGER_JSON_NAME="${ver%/*}.json"
+  curl -fs ${SWAGGER_API_PATH}${SUBPATH} > ${SWAGGER_ROOT_DIR}/${SWAGGER_JSON_NAME}
```
Comment (krousey) [inline hack/after-build/update-swagger-spec.sh]: Double quote.

File: hack/gen-swagger-doc/gen-swagger-docs.sh [hunk 4]
```diff
@@ -41,20 +42,24 @@ cp /swagger-source/"$1".json input.json
...
+# check if the top level models exist in the definitions.adoc. If they exist,
+# their name will be <version>.<model_name>
+VERSION="${1#*_}"
 for m in $top_level_models
 do
-  if grep -xq "=== $1.$m" ./definitions.adoc
+  if grep -xq "=== ${VERSION}.$m" ./definitions.adoc
   then
-    buf+="* <<$1."$m">>\n"
+    buf+="* <<${VERSION}."$m">>\n"
```
Comment (krousey) [inline hack/gen-swagger-doc/gen-swagger-docs.sh]: `$m` should probably be in the quotes here.

File: hack/lib/util.sh [hunk 5]
```diff
@@ -297,4 +297,19 @@ kube::util::group-version-to-pkg-path() {
...
+# default behavior: extensions/v1beta1 -> extensions_v1beta1
+# special case for v1: v1 -> v1
+kube::util::gv-to-swagger-name() {
+  local group_version="$1"
+  case "${group_version}" in
+    v1)
+      echo "v1"
+      ;;
+    *)
+      echo "${ver%/*}_${ver#*/}"
```
Comment (krousey) [inline hack/lib/util.sh]: What is ver? Did you mean group_version?

Comment (caesarxuchao) [inline hack/lib/util.sh]: Oops, right.

**Comment by caesarxuchao** [comment]
@krousey, comments addressed. PTAL. Thanks.

File: hack/update-api-reference-docs.sh [hunk 6]
```diff
@@ -40,12 +40,13 @@ OUTPUT_TMP="${KUBE_ROOT}/${TMP_SUBPATH}"
 
 echo "Generating api reference docs at ${OUTPUT_TMP}"
 
-V1_TMP_IN_HOST="${OUTPUT_TMP_IN_HOST}/v1/"
-V1_TMP="${OUTPUT_TMP}/v1/"
-mkdir -p ${V1_TMP}
-V1BETA1_TMP_IN_HOST="${OUTPUT_TMP_IN_HOST}/extensions/v1beta1/"
-V1BETA1_TMP="${OUTPUT_TMP}/extensions/v1beta1/"
-mkdir -p ${V1BETA1_TMP}
+#DEFAULT_GROUP_VERSIONS="v1 authorization/v1beta1 extensions/v1beta1 componentconfig/v1alpha1 metrics/v1alpha1"
```
Comment (krousey) [inline hack/update-api-reference-docs.sh]: delete comment?

File: hack/update-api-reference-docs.sh [hunk 7]
```diff
@@ -56,13 +57,23 @@ if [[ $(uname) == "Darwin" ]]; then
...
-    https://raw.githubusercontent.com/kubernetes/kubernetes/master/pkg/apis/extensions/v1beta1/register.go
+for ver in $VERSIONS; do
+  TMP_IN_HOST="${OUTPUT_TMP_IN_HOST}/${ver}"
+  REGISTER_FILE_URL="https://raw.githubusercontent.com/kubernetes/kubernetes/master/pkg"
+  if [[ ${ver} == "v1" ]]; then
+    REGISTER_FILE_URL="${REGISTER_FILE_URL}/api/${ver}/register.go"
+  else
+    REGISTER_FILE_URL="${REGISTER_FILE_URL}/apis/${ver}/register.go"
+  fi
+  SWAGGER_JSON_NAME="$(kube::util::gv-to-swagger-name ${ver})"
```
Comment (krousey) [inline hack/update-api-reference-docs.sh]: should quote the nested `${ver}` too

File: hack/update-api-reference-docs.sh [hunk 8]
```diff
@@ -56,13 +57,23 @@ if [[ $(uname) == "Darwin" ]]; then
...
+
+  docker run ${user_flags} \
+    --rm -v ${TMP_IN_HOST}:/output:z \
+    -v ${SWAGGER_PATH}:/swagger-source:z \
+    gcr.io/google_containers/gen-swagger-docs:v5 \
+    ${SWAGGER_JSON_NAME} \
+    ${REGISTER_FILE_URL}
```
Comment (krousey) [inline hack/update-api-reference-docs.sh]: Quotes all around

File: hack/update-api-reference-docs.sh [hunk 9]
```diff
@@ -56,13 +56,23 @@ if [[ $(uname) == "Darwin" ]]; then
...
+    REGISTER_FILE_URL="${REGISTER_FILE_URL}/apis/${ver}/register.go"
+  fi
+  SWAGGER_JSON_NAME="$(kube::util::gv-to-swagger-name "${ver}")"
 
-docker run ${user_flags} --rm -v $V1BETA1_TMP_IN_HOST:/output:z -v ${SWAGGER_PATH}:/swagger-source:z gcr.io/google_containers/gen-swagger-docs:v4.2 \
-    v1beta1 \
-    https://raw.githubusercontent.com/kubernetes/kubernetes/master/pkg/apis/extensions/v1beta1/register.go
+  docker run ${user_flags} \
+    --rm -v "${TMP_IN_HOST}":/output:z \
+    -v "${SWAGGER_PATH}":/swagger-source:z \
```
Comment (caesarxuchao) [inline hack/update-api-reference-docs.sh]: I think using double quotes here is correct, right?

**Comment by caesarxuchao** [comment]
Addressed the comments and asked one question. PTAL. Thanks.

**Comment by krousey** [comment]
LGTM squash and apply the label when you're done.

**Comment by caesarxuchao** [comment]
Squashed. Apply the LGTM. Thanks.

**Pull request merged**